### PR TITLE
Bundler support in bazel

### PR DESCRIPTION
**Pull request by elliottt**
This adds support for managing gem dependencies through bazel.

### Motivation
This is part of an ongoing effort to migrate the snapshot tests to bazel.

### Test plan
There's a single smoke test in the `@gems` package, that can be run with `bazel test @gems//...`. This test is run along with the `@ruby_2_4_3` tests on mac during `test-static-sanitized.sh`.

### CODE REVIEW

File: WORKSPACE [hunk 1]
```diff
@@ -30,6 +30,36 @@ load("@io_bazel_rules_bison//bison:bison.bzl", "bison_register_toolchains")
 
 bison_register_toolchains()
 
+load("//third_party/gems:rules.bzl", "gemfile_lock_deps")
+
+gemfile_lock_deps(
+    name = "gems",
+    gemfile_locks = [
+        "//gems/sorbet/test/snapshot:{}/src/Gemfile.lock".format(test)
+        for test in [
```
Comment (DarkDimius) [inline WORKSPACE]: this is temporary, right? 
can we leave a TODO?

File: third_party/gems/rules.bzl [hunk 2]
```diff
@@ -0,0 +1,249 @@
...
+    last_slash = name.rfind("/")
+    if last_slash >= 0:
+        base = "{}/{}".format(base, name[0:last_slash])
+
+    return base
+
+def _read_file(repo_ctx, label):
+    """
+    Bazel-0.21.0 lacks `repo_ctx.read`, so this simulates it by briefly
+    symlinking the file into the current repo, cat-ing it, and then removing
```
Comment (DarkDimius) [inline third_party/gems/rules.bzl]: 🤦‍♂

File: third_party/gems/rules.bzl [hunk 3]
```diff
@@ -0,0 +1,249 @@
...
+    the symlink.
+    """
+
+    if repo_ctx.attr.debug:
+        print("Reading {}".format(label))
+
+    repo_ctx.symlink(label, "tmp")
```
Comment (DarkDimius) [inline third_party/gems/rules.bzl]: what are isolation guarantees here. Can this race with other version of this command?

File: WORKSPACE (hunk 1 above)
Comment (elliottt) [reply WORKSPACE]: Ultimately it would be good to discover the `Gemfile.lock` files, but I'm not sure how to search the project during the loading phase. I'll add a TODO to this effect.

File: third_party/gems/rules.bzl (hunk 2 above)
Comment (elliottt) [reply third_party/gems/rules.bzl]: Yeah, seriously.

**Comment by DarkDimius** [comment]
> gems/sorbet/test/snapshot/BUILD

what is this empty file for?

File: third_party/gems/bundle-env.sh [hunk 4]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
+
```
Comment (DarkDimius) [inline third_party/gems/bundle-env.sh]: Can we add a comment explaining what this script is supposed to take & do?

File: third_party/gems/bundle.sh [hunk 5]
```diff
@@ -0,0 +1,32 @@
+#!/bin/bash
+
```
Comment (DarkDimius) [inline third_party/gems/bundle.sh]: ditto

File: third_party/gems/rules.bzl (hunk 3 above)
Comment (elliottt) [reply third_party/gems/rules.bzl]: I think that there's no race here, as it's only used to sequentially read the `Gemfile.lock` files passed in via `gemfile_locks`.

File: third_party/gems/rules.bzl [hunk 6]
```diff
@@ -0,0 +1,249 @@
...
+    if sha256 == None:
+        sha256 = ""
+
+    result = repo_ctx.download(output = output, url = url, sha256 = sha256)
+
+    return result.sha256
+
+BUNDLER_VERSIONS = {
+    "2.0.1": {
+        "sha256": "",
```
Comment (DarkDimius) [inline third_party/gems/rules.bzl]: ?

File: third_party/gems/test/smoke_test.sh [hunk 7]
```diff
@@ -0,0 +1,57 @@
+#!/bin/bash
+
```
Comment (DarkDimius) [inline third_party/gems/test/smoke_test.sh]: ditto

File: third_party/gems/bundle.sh [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+# --- begin runfiles.bash initialization ---
+# Copy-pasted from Bazel's Bash runfiles library (tools/bash/runfiles/runfiles.bash).
+set -euo pipefail
+if [[ ! -d "${RUNFILES_DIR:-/dev/null}" && ! -f "${RUNFILES_MANIFEST_FILE:-/dev/null}" ]]; then
```
Comment (DarkDimius) [inline third_party/gems/bundle.sh]: rather than being copy-pasted into every script, can it be reused? 
can we `source` something?

File: third_party/gems/rules.bzl (hunk 6 above)
Comment (elliottt) [reply third_party/gems/rules.bzl]: Thanks for catching this!

File: third_party/gems/vendor_cache.BUILD [hunk 9]
```diff
@@ -0,0 +1,14 @@
+
+filegroup(
+    name = "cache",
+    srcs = [ %{gemfiles} ],
```
Comment (DarkDimius) [inline third_party/gems/vendor_cache.BUILD]: I didn't understand what this does, could you leave comments?

File: third_party/gems/bundle.sh (hunk 8 above)
Comment (elliottt) [reply third_party/gems/bundle.sh]: Unfortunately, this is what sets up `rlocation` so that we can reliably find artifacts from `data` attributes.

File: third_party/gems/vendor_cache.BUILD (hunk 9 above)
Comment (elliottt) [reply third_party/gems/vendor_cache.BUILD]: I should have removed this template, thanks for catching this.

File: .buildkite/test-static-sanitized.sh [hunk 10]
```diff
@@ -12,10 +12,10 @@ esac
 
 if [[ "linux" == "$platform" ]]; then
   CONFIG_OPTS="--config=buildfarm-sanitized-linux"
-  RUBY_TESTS=""
+  MAC_TESTS=""
```
Comment (DarkDimius) [inline .buildkite/test-static-sanitized.sh]: We've recently changed mac tests to only run in master.
This isn't going to be tested in branches.

How long do you expect it to be only tested on mac? should we re-enable branch tests on mac?

Comment (elliottt) [reply .buildkite/test-static-sanitized.sh]: We could re-enable the branch tests on mac for now, moving these tests to linux will just require fixing that issue with `probes.h` on the ruby build.

File: third_party/gems/bundle.sh (hunk 8 above)
Comment (elliottt) [reply third_party/gems/bundle.sh]: (Copy-pasting that snippet is the official recommendation from bazel: https://github.com/bazelbuild/bazel/blob/master/tools/bash/runfiles/runfiles.bash)

File: WORKSPACE (hunk 1 above)
Comment (jez) [reply WORKSPACE]: Isn't this going to be a big footgun? What will be the signal that I need to change this list?

I'd love to figure out if we can discover these with some sort of glob.

File: third_party/gems/test/smoke_test.sh (hunk 7 above)
Comment (elliottt) [reply third_party/gems/test/smoke_test.sh]: I've added a comment.

File: third_party/gems/bundle.sh (hunk 5 above)
Comment (elliottt) [reply third_party/gems/bundle.sh]: I've added a comment.

File: third_party/gems/bundle-env.sh (hunk 4 above)
Comment (elliottt) [reply third_party/gems/bundle-env.sh]: I've added a comment.

File: third_party/gems/BUILD [hunk 11]
```diff
@@ -0,0 +1 @@
+licenses(["notice"])
```
Comment (jez) [inline third_party/gems/BUILD]: What does this have to do with the rest of the PR?

Comment (elliottt) [reply third_party/gems/BUILD]: Absolutely nothing. Bazel requires files in third_party to have a BUILD file with a `licenses` declaration.

File: third_party/gems/bundle-env.sh [hunk 12]
```diff
@@ -0,0 +1,34 @@
...
+# --- begin runfiles.bash initialization ---
+# Copy-pasted from Bazel's Bash runfiles library (tools/bash/runfiles/runfiles.bash).
+set -euo pipefail
```
Comment (jez) [inline third_party/gems/bundle-env.sh]: This affects the script globally. Even if it was copy / pasted, can we move it to the top of the file?

File: third_party/gems/rules.bzl (hunk 3 above)
Comment (DarkDimius) [reply third_party/gems/rules.bzl]: 👍

File: WORKSPACE (hunk 1 above)
Comment (elliottt) [reply WORKSPACE]: `glob` isn't available during the loading phase, so we'll have to look into ways to do this portably from a script.

File: third_party/gems/bundle-env.sh [hunk 13]
```diff
@@ -0,0 +1,34 @@
+#!/bin/bash
+
+# --- begin runfiles.bash initialization ---
+# Copy-pasted from Bazel's Bash runfiles library (tools/bash/runfiles/runfiles.bash).
```
Comment (jez) [inline third_party/gems/bundle-env.sh]: Ultimately what's the purpose of this snippet?

My guess is that it's basically just here to allow for `rlocation` to be a function that people can call, is that right?

Maybe we should add a link here:

https://github.com/bazelbuild/bazel/blob/master/tools/bash/runfiles/runfiles.bash

at the top, or better yet link to a commit sha where you copied this snippet from?

File: third_party/gems/bundle-env.sh (hunk 12 above)
Comment (elliottt) [reply third_party/gems/bundle-env.sh]: Happy to move that part out of the region copied from bazel. I had left it inside to ensure that it was consistent with what the bazel docs were recommending (https://github.com/bazelbuild/bazel/blob/master/tools/bash/runfiles/runfiles.bash).

File: third_party/gems/bundle-env.sh (hunk 13 above)
Comment (jez) [reply third_party/gems/bundle-env.sh]: (or if you modified it that's fine too, just explain that)

File: third_party/gems/bundle-env.sh (hunk 12 above)
Comment (elliottt) [reply third_party/gems/bundle-env.sh]: Also, note that version is newer than what we get with bazel-0.21.0. We're using the version that they're referring to as `version 1` in that file.

File: third_party/gems/bundle-env.sh (hunk 13 above)
Comment (elliottt) [reply third_party/gems/bundle-env.sh]: I'm happy to include a link to the original file. It's just there because the environment that your script is run in can be wildly different depending on the context it's run from (another script vs through `bazel run` etc.).

File: third_party/gems/bundle-env.sh (hunk 12 above)
Comment (jez) [reply third_party/gems/bundle-env.sh]: ^ even better to link to a SHA then.

File: .buildkite/test-static-sanitized.sh (hunk 10 above)
Comment (DarkDimius) [reply .buildkite/test-static-sanitized.sh]: Happy to re-enable them for now. Please edit `pipeline.yaml` for that removing branch filter

File: third_party/gems/setup_bundler.sh [hunk 14]
```diff
@@ -0,0 +1,21 @@
+#!/bin/bash
+
+# A template script to unpack the internals of the bundler gem into a local tree
+# that looks like the site_ruby tree that the tracepoint tracer is expecting to
+# find
+
+set -euo pipefail
+
+cd "%{bundler}"
```
Comment (jez) [inline third_party/gems/setup_bundler.sh]: What is this `%{...}` syntax?

Comment (elliottt) [reply third_party/gems/setup_bundler.sh]: It's a template that's replaced during the repository rule. I'm happy to pick a different word, but this seemed like an easy way to pick substitution strings. Here's a use of `repo_ctx.template` where substitutions are defined:
https://github.com/sorbet/sorbet/pull/939/files#diff-d6d62f775ac82d0f76fbca30cb412757R164

File: third_party/gems/bundle-env.sh (hunk 12 above)
Comment (elliottt) [reply third_party/gems/bundle-env.sh]: I've modified all the snippets to refer to the SHA for 0.21.0.

File: third_party/gems/test/smoke_test.sh [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+            "$RUNFILES_MANIFEST_FILE" | cut -d ' ' -f 2-)"
+else
+  echo >&2 "ERROR: cannot find @bazel_tools//tools/bash/runfiles:runfiles.bash"
+  exit 1
+fi
+# --- end runfiles.bash initialization ---
+
+# put bundle in the path
+# shellcheck disable=SC1083
+PATH="$(dirname "$(rlocation %{workspace}/bundler/bundle)"):$PATH"
```
Comment (jez) [inline third_party/gems/test/smoke_test.sh]: Does the `rlocation` fail if the argument doesn't exist in some list of fixtures or something like that?

i.e., why are we not just doing:

```bash
$(rlocation %{workspace}/bundler)
```

and instead wrapping it in a `dirname` call?

Comment (elliottt) [reply third_party/gems/test/smoke_test.sh]: `rlocation` will only work with paths you've specified in the dependencies of the `sh_binary`. Directories can not go in that list, so what I've seen people do is to name a file in that directory, and use `dirname` to get the actual directory of interest.

File: third_party/gems/bundle-env.sh (hunk 13 above)
Comment (elliottt) [reply third_party/gems/bundle-env.sh]: I added links to the version we're using with 0.21.0.

File: third_party/gems/test/smoke_test.sh [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+
+# bundle requires $HOME
+HOME=$PWD
+export HOME
+
+# ensure that bundle works
+bundle --help
+
+
+# shellcheck disable=SC1083
```
Comment (jez) [inline third_party/gems/test/smoke_test.sh]: I think we should disable this error code. Can you audit this PR to find instances where we can replace

```bash
cp "$(rlocation %{workspace}/test/Gemfile)" .
```

with

```bash
cp "$(rlocation "%{workspace}/test/Gemfile")" .
```

? The error it's saying is that "you might have meant to do `echo {1,2}`" which has a special meaning in the shell (expands to `echo 1 2`). We should be able to say we don't care about shell expansion.

File: .buildkite/test-static-sanitized.sh (hunk 10 above)
Comment (elliottt) [reply .buildkite/test-static-sanitized.sh]: I've pushed a change to re-enable it for branches. Once @ruby_2_4_3 works on linux, we can move it back to `master` only.

File: third_party/gems/test/smoke_test.sh (hunk 16 above)
Comment (jez) [reply third_party/gems/test/smoke_test.sh]: ^ edited (never* disable)

Comment (elliottt) [reply third_party/gems/test/smoke_test.sh]: Ah, sorry. I'll go through and remove those.

File: third_party/gems/test/smoke_test.sh [hunk 17]
```diff
@@ -0,0 +1,60 @@
...
+cp "$(rlocation %{workspace}/test/Gemfile.lock)" .
+
+mkdir vendor
+
+# shellcheck disable=SC1083
+VENDOR_CACHE=$(dirname "$(rlocation %{workspace}/test/vendor/cache/cantor-1.2.1.gem)")
+ln -s "$VENDOR_CACHE" vendor/cache
+
+# setup 'cantor'
+bundle install --deployment --local
```
Comment (jez) [inline third_party/gems/test/smoke_test.sh]: What do these two flags do? I've never used them myself before. Is there something special in the way this interacts with bazel?

File: third_party/gems/test/smoke_test.sh (hunk 16 above)
Comment (elliottt) [reply third_party/gems/test/smoke_test.sh]: I've removed all instances of SC1083 being disabled.

File: third_party/gems/test/smoke_test.sh (hunk 17 above)
Comment (elliottt) [reply third_party/gems/test/smoke_test.sh]: `--deployment` causes it to put gems into `vendor/cache`, and `--local` causes it to not pull anything from rubygems.org.

https://bundler.io/v2.0/man/bundle-install.1.html#DEPLOYMENT-MODE

File: WORKSPACE (hunk 1 above)
Comment (jez) [reply WORKSPACE]: ack that's too bad. In that case the TODO seems good enough.

**Comment by elliottt** [comment]
> > gems/sorbet/test/snapshot/BUILD
> 
> what is this empty file for?

You need a BUILD file present to be able to refer to files, so this is what enables the labels in the WORKSPACE file.

File: third_party/gems/gemfile.bzl [hunk 18]
```diff
@@ -0,0 +1,38 @@
+def _parse_package(line):
+    """
+    Parse an exact package specification from a single line of a Gemfile.lock.
+    If the line did not contain an exact specification (something of the form
+    `package (version)`, return None.
+    """
+
+    prefix = line[0:4]
```
Comment (jez) [inline third_party/gems/gemfile.bzl]: I would love to get a comment on these lines that looks like what you expect the string to look like when it's being parsed.

File: third_party/gems/bundler.BUILD [hunk 19]
```diff
@@ -0,0 +1,35 @@
...
+    visibility = [ "//visibility:public" ],
+)
+
+sh_binary(
+    name = "bundle",
+    data = [ ":runtime_deps", ":bundle-env" ],
+    srcs = [ "bundle.sh" ],
+    deps = [ "@bazel_tools//tools/bash/runfiles" ],
+    visibility = [ "//visibility:public" ],
+)
```
Comment (jez) [inline third_party/gems/bundler.BUILD]: What's the difference between this `sh_binary` (bundle) and the one above (bundler)?

File: .buildkite/pipeline.yaml [hunk 20]
```diff
@@ -25,7 +25,7 @@ steps:
 
   - label: ":mac: test-static-sanitized.sh (master only)"
```
Comment (jez) [inline .buildkite/pipeline.yaml]: This is no longer `master only`.

I might have also just deleted the line instead of commenting it out, but that's your call.

File: WORKSPACE [hunk 21]
```diff
@@ -30,6 +30,39 @@ load("@io_bazel_rules_bison//bison:bison.bzl", "bison_register_toolchains")
...
+            "partial/bad-hash",
+            "partial/bad-t",
+            "partial/codecov",
+            "partial/create-config",
+            "partial/explosive-object",
+            "partial/fake-object",
+            "partial/fake-rails",
+            # "partial/local_gem",
```
Comment (jez) [inline WORKSPACE]: What's up with this guy: local_gem?

File: third_party/gems/test/smoke_test.sh (hunk 15 above)
Comment (jez) [reply third_party/gems/test/smoke_test.sh]: :derpbazel:

File: third_party/gems/setup_bundler.sh (hunk 14 above)
Comment (jez) [reply third_party/gems/setup_bundler.sh]: Ah interesting I didn't realize we got to choose the substitution characters.

I might have gone with `{{bundler}}` myself because that's what a lot of other templating libraries use, but that's just my thoughts and not blocking.

**Comment by jez** [comment]
I still don't have a good mental model for how this might break or what parts I expect to be brittle, but I think at a surface level I've left all my questions of syntax.

Can I get a gut check from you on how you feel (good / bad / skeptical / confident / other) about this change? If you expect that it was going to break somewhere, where would that be?

File: third_party/gems/bundler.BUILD (hunk 19 above)
Comment (elliottt) [reply third_party/gems/bundler.BUILD]: There is no difference, this is just to preserve the expected interface to bundler, which is that both the `bundle` and `bundler` executables exist.

File: .buildkite/pipeline.yaml (hunk 20 above)
Comment (elliottt) [reply .buildkite/pipeline.yaml]: Yes, I made a comment indicating that I was doing this. I'm happy to remove the line.

File: WORKSPACE (hunk 21 above)
Comment (elliottt) [reply WORKSPACE]: The local gem causes the build to fail, as the gem does not exist on rubygems.org. I think that it's possible to handle this case better, but I wanted to get this work merged so that it didn't continue to sit on a branch. I'll add a comment indicating what I think needs to be done to support this.

File: third_party/gems/setup_bundler.sh (hunk 14 above)
Comment (elliottt) [reply third_party/gems/setup_bundler.sh]: I'm happy to change this.

Comment (elliottt) [reply third_party/gems/setup_bundler.sh]: I've changed them to `{{name}}`.

File: third_party/gems/gemfile.bzl (hunk 18 above)
Comment (elliottt) [reply third_party/gems/gemfile.bzl]: There is a description in backticks in the comment, but I'll add a more thorough example.

Comment (elliottt) [reply third_party/gems/gemfile.bzl]: I've expanded the comment, please take another read through.

File: .buildkite/pipeline.yaml (hunk 20 above)
Comment (elliottt) [reply .buildkite/pipeline.yaml]: I've removed the line.

**Comment by DarkDimius** [comment]
> You need a BUILD file present to be able to refer to files, so this is what enables the labels in the WORKSPACE file.

could we leave this as a comment in that file?

File: WORKSPACE (hunk 21 above)
Comment (elliottt) [reply WORKSPACE]: I've fixed this by only looking for gem names in the `GEM` section.
